### PR TITLE
Add a debug API endpoint to quit users who have been idle for a while

### LINKED ISSUE #388: Please can we regularly and automatically reap idle-presence connections on all networks
e.g. Snoonet IRC users are enraged at the amount of join/part spam when the bridge farts

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #388 

This PR introduces the ability to periodically remove users who have not interacted with the bridge (or homeserver) in a given period. It uses [matrix-lastactive](https://github.com/Half-Shot/matrix-lastactive) in order to detect which users are offline.

This is checked first by local activity (e.g. sending messages/member events), then by presence (if enabled) and finally by use of the admin api for local users (if the user is allowed to make admin calls).

This fundamentally automates the process that matrix.org is already using for reaping users on it's Freenode bridge.

### CODE REVIEW

**Comment by Half-Shot** [comment]
@turt2live I've given you the gift of ✔️

File: lib/bridge/IrcBridge.js [hunk 1]
```diff
@@ -46,6 +48,14 @@ function IrcBridge(config, registration) {
...
     };
     this.joinedRoomList = [];
+    const connectionReapingEnabled = config.ircService.connectionReaper &&
+        config.ircService.connectionReaper.enabled;
+    this.activityTracker = connectionReapingEnabled ? new MatrixActivityTracker(
+        this.config.homeserver.url,
+        registration.as_token,
+        this.config.homeserver.domain,
+        this.config.homeserver.enablePresence,
+    ) : null;
```
Comment (turt2live) [inline lib/bridge/IrcBridge.js]: wouldn't you want to specify a non-silent logger?

File: lib/bridge/IrcBridge.js [hunk 2]
```diff
@@ -441,6 +451,34 @@ IrcBridge.prototype.run = Promise.coroutine(function*(port) {
         server.config.ircClients.reconnectIntervalMs = reconnectIntervalsMap[server.domain];
     });
 
+    if (this.activityTracker) {
+        const THIRTY_DAYS = 30 * 24 * 60 * 60 * 1000;
+        const MAX_DELAY = 2147483647;
```
Comment (turt2live) [inline lib/bridge/IrcBridge.js]: Any reason for specifically 24.85513 days?

File: config.sample.yaml [hunk 3]
```diff
@@ -460,6 +460,11 @@ ircService:
     # "off" - Defaults to disabled, users can choose to enable.
     # "force-off" - Disabled, cannot be enabled.
     mapIrcMentionsToMatrix: "on" # This can be "on", "off", "force-off".
+  
+    connectionReaper:
```
Comment (turt2live) [inline config.sample.yaml]: docs on what this is would be great

File: lib/bridge/IrcBridge.js [hunk 4]
```diff
@@ -441,6 +451,34 @@ IrcBridge.prototype.run = Promise.coroutine(function*(port) {
...
+        if (!this.config.ircService.connectionReaper.maxIdleTime) {
+            this.config.ircService.connectionReaper.maxIdleTime = THIRTY_DAYS;
```
Comment (turt2live) [inline lib/bridge/IrcBridge.js]: THIRTY_DAYS is larger than your MAX_DELAY, so it'll always get capped later on.

**Comment by Half-Shot** [comment]
>also test failures

say what?

**Comment by turt2live** [comment]
I swear it was red when I started reviewing it

**Comment by Half-Shot** [comment]
(sorry for the force push, I thought I hadn't pushed before)

File: config.sample.yaml [hunk 5]
```diff
@@ -460,7 +460,6 @@ ircService:
     # "off" - Defaults to disabled, users can choose to enable.
     # "force-off" - Disabled, cannot be enabled.
     mapIrcMentionsToMatrix: "on" # This can be "on", "off", "force-off".
-
```
Comment (turt2live) [inline config.sample.yaml]: irrelevant diff

**Comment by Half-Shot** [comment]
>lgtm assuming the logging code outputs a compatible logger

I checked, it does :).

>Also if we have documentation on the debug API, we should put this endpoint in there.

We don't, but I'll add a wiki page.

**Pull request merged**